Redesigned, more detailed and updated README with product-focused approach (#182)
roscisz/TensorHive: README.md

- :ocean: Other popular tools are simply an overkill, have different purpose or require a lot of time to spend on reading documentation, installation and configuration (Graphana, Kubernetes, Slurm)
- :penguin: People that are using your infrastructure expect only one interface for all the things related to training models (besides terminal): monitoring, reservation calendar and scheduling distributed jobs 
- :collision: Can't risk messing up sensitive configuration by installing software on each individual machine, prefering centralized solution which can be managed from one place.

2. You're a **standalone user** who has access to beefy GPUs scatterd across multiple machines.
- :part_alternation_mark: You want to be able to determine if batch size is too small or if there's a bottleneck when moving data from memory to GPU - charts with metrics such as `gpu_util`, `mem_util`, `mem_used` are great for this purpose
- :date: Visualizing names of training experiments using calendar helps you track how you're progressing on the project
- :snake: Launching distributed trainings is essential for you, no matter what the framework is

> - :dizzy_face: Managing a list of training commands for all your distributed training experiments drives you nuts (because you didn't wrap them into bash scripts)

Author: yes, it's kind of ambiguous. Should I delete that whole content inside brackets or replace with something specific?

> - :dizzy_face: Managing a list of training commands for all your distributed training experiments drives you nuts (because you didn't wrap them into bash scripts)

Reviewer: Maybe delete the comment in the brackets?

> - :dizzy_face: Managing a list of training commands for all your distributed training experiments drives you nuts (because you didn't wrap them into bash scripts)

Author: Done

- :zzz: Remembering to manually launch the training before going sleep is no fun anymore

What TensorHive has to offer
-----------------------------
:zero: Dead-simple one-machine installation and configuration, no `sudo` requirements

:one: Users can make GPU reservations for specific time range in advance via **reservation mechanism**

&nbsp;&nbsp;&nbsp;&nbsp; :arrow_right: no more frustration caused by rules: **"first come, first served"** or **"the law of the jungle"**.

:two: Users can prepare and schedule custom tasks (commands) to be run on selected GPUs and hosts

&nbsp;&nbsp;&nbsp;&nbsp; :arrow_right: automate and simplify **distributed trainings** - **"one button to rule them all"**

:three: Gather all useful GPU metrics, from all configured hosts **in one dashboard**

> &nbsp;&nbsp;&nbsp;&nbsp; :arrow_right: no more manually logging in to each individual machine in order to check if GPU is currently taken or not

Author: Yes, taken is not the best word, "utilized" is much better, thumbs up. Hmmm and what about "in use"?

> &nbsp;&nbsp;&nbsp;&nbsp; :arrow_right: no more manually logging in to each individual machine in order to check if GPU is currently taken or not

Reviewer: OK, changed to "in use" and commited.

For more details, check out the [full list of features](#features)

Getting started
---------------
### Prerequisites

> * All hosts must be accessible via SSH, without password, using SSH Key-Based Authentication ([How to set up SSH keys](https://www.shellhacks.com/ssh-login-without-password/)

Author: Is there a chance that some has machines which can be accessed by anyone (literally) also without any password - truly bad idea. I know that in some very simple, local scenarios this may be a thing, but maybe u're right...

> * All hosts must be accessible via SSH, without password, using SSH Key-Based Authentication ([How to set up SSH keys](https://www.shellhacks.com/ssh-login-without-password/)

Reviewer: There can be another auth method used, like for example host-based.

    - [x] Metrics and active processes
    - [ ] Detailed harware specification 
- [x] :calendar: Calendar view
    - [x] Allow making reservations for selected GPUs
    - [x] Edit reservations
    - [x] Cancel reservations
    - [ ] Attach jobs to reservation
- [x] :baby_symbol: Task nursery
    - [x] Create parametrized tasks and assign to hosts, automatically set `CUDA_VISIBLE_DEVICES`
    - [x] Buttons for task spawning/scheduling/termination/killing actions
    - [x] Fetch log produced by running task
    - [x] Group actions (spawn, schedule, terminate, kill selected)
- [ ] :straight_ruler: Detailed hardware specification panel (CPU clock speed, RAM, etc.)
- [ ] :penguin: Admin panel
    - [ ] User banning
    - [ ] Accept/reject reservation requests
    - [ ] Modify rules on-the-fly (without restarting)
    - [ ] Show popups to users (something like message of the day - `motd`)
    
#### CLI
- [ ] Implement command-line app that communicates with core via API
- [ ] Migrate all features from web app that don't require GUI (so no charts)

#### API
- [x] OpenAPI 2.0 specification with Swagger UI
- [x] User authentication via JWT

> Deployment in production (for admins)

Reviewer: ```suggestion
Production deployment (for admins)
```

Author: Yes, "Deployment in production" is grammatically wrong. 
What about "Deployment to production"? (which one we choose then)

TensorHive
===
![](https://img.shields.io/badge/release-v0.3-brightgreen.svg?style=popout-square)
![](https://img.shields.io/badge/pypi-v0.3-brightgreen.svg?style=popout-square)
![](https://img.shields.io/badge/Issues%20and%20PRs-welcome-yellow.svg?style=popout-square)
![](https://img.shields.io/badge/platform-Linux-blue.svg?style=popout-square)
![](https://img.shields.io/badge/hardware-Nvidia-green.svg?style=popout-square)
![](https://img.shields.io/badge/python-3.5%20|%203.6%20|%203.7-blue.svg?style=popout-square)
![](https://img.shields.io/badge/license-Apache%202.0-blue.svg?style=popout-square)

<img src="https://github.com/roscisz/TensorHive/raw/master/images/logo_small.png" height="130" align="left">

TensorHive is an open source system for monitoring and managing computing resources across multiple hosts.
It solves the most common problems and nightmares about accessing and sharing your AI-oriented infrastructure across multiple, often competing users.

It's designed with __simplicty, flexibility and configuration-friendliness__ in mind.

> About project

Reviewer: ```suggestion
Use cases
```

- :penguin: People using your infrastructure expect only one interface for all the things related to training models (besides terminal): monitoring, reservation calendar and scheduling distributed jobs 
- :collision: Can't risk messing up sensitive configuration by installing software on each individual machine, prefering centralized solution which can be managed from one place.

2. You're a **standalone user** who has access to beefy GPUs scatterd across multiple machines.
- :part_alternation_mark: You want to be able to determine if batch size is too small or if there's a bottleneck when moving data from memory to GPU - charts with metrics such as `gpu_util`, `mem_util`, `mem_used` are great for this purpose
- :date: Visualizing names of training experiments using calendar helps you track how you're progressing on the project
- :snake: Launching distributed trainings is essential for you, no matter what the framework is
- :dizzy_face: Managing a list of training commands for all your distributed training experiments drives you nuts (because you didn't wrap them into bash scripts)

> - :zzz: Remembering to manually launch the training before going sleep is no fun anymore

Reviewer: ```suggestion
- :zzz: Remembering to manually launch the training before going sleep is no fun any more.
```

Author: ![image](https://user-images.githubusercontent.com/12485656/61938116-9dda6d80-af90-11e9-94e7-6388a7ca0aed.png)

Reviewer: OK, thx. I'll leave the dot though :)

    - [x] :warning:	Send warning messages to terminal of users who violate the rules
    - [x] :mailbox_with_no_mail: Send e-mail warnings
    - [ ] :bomb: Kill unwanted processes
- [X] :rocket: Task nursery and scheduling
    - [x] :old_key: Execute any command in the name of a user
    - [x] :alarm_clock: Schedule spawn and termination
    - [x] :repeat: Synchronize process status
    - [x] :factory: Use `screen` command as backend - user can easily attach to running task
    - [x] :skull: Remote process interruption, termination and kill
    - [x] :floppy_disk: Save stdout to disk
    - [ ] :page_facing_up: Capture stderr
- [x] :watch: Track wasted (idle) time during reservation
    - [x] :hocho: Gather and calculate average gpu and mem utilization
    - [ ] :loudspeaker: Remind user when his reservation starts and ends
    - [ ] :incoming_envelope: Send e-mail if idle for too long
    
#### Web
- [x] :chart_with_downwards_trend: Configurable charts view
    - [x] Metrics and active processes
    - [ ] Detailed harware specification 
- [x] :calendar: Calendar view
    - [x] Allow making reservations for selected GPUs
    - [x] Edit reservations
    - [x] Cancel reservations

>     - [ ] Attach jobs to reservation

Reviewer: ```suggestion
    - [ x] Attach jobs to reservation
```

- [x] :baby_symbol: Task nursery
    - [x] Create parametrized tasks and assign to hosts, automatically set `CUDA_VISIBLE_DEVICES`
    - [x] Buttons for task spawning/scheduling/termination/killing actions
    - [x] Fetch log produced by running task
    - [x] Group actions (spawn, schedule, terminate, kill selected)
- [ ] :straight_ruler: Detailed hardware specification panel (CPU clock speed, RAM, etc.)
- [ ] :penguin: Admin panel
    - [ ] User banning
    - [ ] Accept/reject reservation requests
    - [ ] Modify rules on-the-fly (without restarting)
    - [ ] Show popups to users (something like message of the day - `motd`)
    
#### CLI
- [ ] Implement command-line app that communicates with core via API
- [ ] Migrate all features from web app that don't require GUI (so no charts)

#### API
- [x] OpenAPI 2.0 specification with Swagger UI
- [x] User authentication via JWT

> TensorHive is currently being used on production in the following environments:

Reviewer: ```suggestion
TensorHive is currently being used in production in the following environments:
```

| ![](https://cdn.pg.edu.pl/ekontakt-updated-theme/images/favicon/favicon-16x16.png?v=jw6lLb8YQ4) [Lab at GUT](https://eti.pg.edu.pl/katedra-architektury-systemow-komputerowych/main) | 18x machines with GTX 1060 6GB | 20+ |
| ![](http://martyniak.tech/images/gradient_logo_small-628ed211.png)[Gradient PG](http://gradient.eti.pg.gda.pl/en/) | TITAN X 12GB | 10+ |
| ![](https://res-4.cloudinary.com/crunchbase-production/image/upload/c_lpad,h_20,w_20,f_auto,q_auto:eco/v1444894092/jeuh0l6opc159e1ltzky.png) [VoiceLab - Conversational Intelligence](https://www.voicelab.ai) | 30+ GTX and RTX cards | 10+

Application examples and benchmarks
--------
Along with TensorHive, we are developing a set of [**sample deep neural network training applications**](https://github.com/roscisz/TensorHive/tree/master/examples) in Distributed TensorFlow which will be used as test applications for the system. They can also serve as benchmarks for single GPU, distributed multi-GPU and distributed multi-node architectures. For each example, a full set of instructions to reproduce is provided.

<hr/>

TensorHive architecture (simplified)
-----------------------

> This diagram will help you grasp the rough concept of the system.

Reviewer: ```suggestion
This diagram will help you to grasp the rough concept of the system.
```
